src/blockchain/compact_filters/p2p.rs

```diff
@@ -0,0 +1,539 @@
+                "dnsseed.bitcoin.dashjr.org",
+                "seed.bitcoinstats.com",
+                "seed.bitcoin.jonasschnelli.ch",
+                "seed.btc.petertodd.org",
+                "seed.bitcoin.sprovoost.nl",
+                "dnsseed.emzy.de",
+                "seed.bitcoin.wiz.biz",
+            ],
+            Network::Testnet => &[
+                "testnet-seed.bitcoin.jonasschnelli.ch",
+                "seed.tbtc.petertodd.org",
+                "seed.testnet.bitcoin.sprovoost.nl",
+                "testnet-seed.bluematt.me",
+            ],
+            Network::Regtest => &[],
+            Network::Signet => &[],
+        };
+
+        for seedhost in seedhosts.iter() {
+            if let Ok(lookup) = (*seedhost, port).to_socket_addrs() {
+                for host in lookup {
+                    if host.is_ipv4() {
+                        seeds.push_back(host);
+                    }
+                }
+            }
+        }
+        seeds
+    }
+}
+
+/// Crawler structure that will interface with Discovery and public bitcoin network
+///
+/// Address manager will spawn multiple crawlers in separate threads to discover new addresses.
+struct AddressWorker {
+    discovery: Arc<Mutex<AddressDiscovery>>,
+    sender: Sender<AddressMessage>,
+    network: Network,
+}
+
+impl AddressWorker {
+    fn new(
+        discovery: Arc<Mutex<AddressDiscovery>>,
+        sender: Sender<AddressMessage>,
+        network: Network,
+    ) -> AddressWorker {
+        AddressWorker {
+            discovery,
+            sender,
+            network,
+        }
+    }
+
+    fn try_receive_addr(&mut self, peer: &Peer) {
+        if let Some(NetworkMessage::Addr(new_addresses)) =
+            peer.recv("addr", Some(Duration::from_secs(1))).unwrap()
+        {
+            self.consume_addr(new_addresses);
+        }
+    }
+
+    fn consume_addr(&mut self, addrs: Vec<(u32, Address)>) {
+        for (_, addr) in addrs.iter() {
+            if let Ok(socket_addr) = addr.socket_addr() {
+                self.sender
+                    .send(AddressMessage::NewPending(socket_addr))
+                    .unwrap();
+            }
+        }
+    }
+
+    fn work(&mut self) {
+        loop {
+            let next_address = {
+                let mut address_discovery = self.discovery.lock().unwrap();
+                address_discovery.get_next()
+            };
+
+            match next_address {
+                Some(address) => {
+                    let potential_peer = Peer::connect_with_timeout(
+                        address,
+                        Duration::from_secs(3),
+                        Arc::new(Mempool::default()),
+                        self.network,
+                    );
+
+                    if let Ok(peer) = potential_peer {
+                        peer.send(NetworkMessage::GetAddr).unwrap();
+                        self.try_receive_addr(&peer);
+                        self.try_receive_addr(&peer);
```

Comment:
I guess it was to give it one more chance. I removed it in the last changes. But I am not fully sure that it doesn't have any impact.

I am having trouble syncing from the existing seed. New addresses are not updating in the pending list. But it works with a previously populated cache.

It could be related to this. Will investigate more.

Reply:
A peer can respond to get addresses in multiple replies and the current peer code ever only expects one reply.  We should probably at some point refactor some of the peer communication layer to be a bit more dynamic but until then I thought it might be useful to poll twice for addr replies.  

I think I logged it at some point and did see it receiving addresses on the second try so it might be helpful.

Reply:
Yes makes sense. At one point I have seen (without the second try_receive) that new spending is not being updated in the discovery. Or it's getting updated with just 1 new address. As a result, the discovery pending list is being quickly exhausted and the workers get stuck.

I changed it back to double try and it didn't happen after that. But in any case, we should handle that situation properly. What to do when the pending list depletes in the discovery and workers don't have any more addresses to try. This can happen for multiple other reasons too, ex: A worker got new addresses and waiting for the write lock. In the meantime, other workers deplete the pending list and then nobody has any new addresses to try and they all wait.

To mitigate part of this I changed the pending addition task from the manager to worker. So it's of one thread trying to add pending to discovery it's all the worker thread and hopefully, someone will get to add new pending before the list depletes.

It's still nasty if that happens. There is no run time error, everything just pauses, and the user doesn't have any clue what went wrong.

The above refactoring removed the need for `AddressMessages` as it's just only one type of message the worker will send to the manager.